magma/compile_guard.py

```diff
@@ -55,9 +56,51 @@ def pre_finalize(self):
 
     def _process_instance(self, inst):
         for port in inst.interface.inputs(include_clocks=True):
-            self._process_port(port)
-
-    def _process_port(self, port, ref=None, kwargs=None):
+            self._process_input(port)
+        for port in inst.interface.outputs():
+            self._process_output(port)
+
+    def _get_flat_drivees(self, port):
+        if isinstance(port, Array):
+            return sum([self._get_flat_drivees(x) for x in port], [])
+        if isinstance(port, Tuple):
+            return sum([self._get_flat_drivees(x) for x in port.values()], [])
+        return port.driving()
+
+    def _get_top_ref(self, ref):
+        if isinstance(ref, TupleRef):
+            return self._get_top_ref(ref.tuple.name)
+        if isinstance(ref, ArrayRef):
+            return self._get_top_ref(ref.array.name)
+        return ref
+
+    def _is_external(self, value):
+        top_ref = self._get_top_ref(value.name)
+        if isinstance(top_ref, DefnRef):
+            return True
+        if isinstance(top_ref, InstRef):
+            return top_ref.inst not in self._instances
+        return False
```

Comment:
This doesn't handle anonymous values created outside the compile-guarded context right? Fine if so, should just add a comment here.

(e.g.
```python
x = Bit()
with compile_guard(...):
    inst = Ckt()
    x @= inst.O  # _is_external(x) == False, even though it is external.
```
)

But perhaps in that case it doesn't really matter?

Reply:
I might feel better if we `raise NotImplemented()` instead of returning `False` incorrectly. I imagine we can say that's a case we explicitly don't handle, rather than just handling it wrong.

Reply:
It works with anonymous values created inside the compile guard context that only refer to internal instances, so if we raise an error then we remove that use case.

Reply:
Actually it seems that the input anons are raising an error, so might as well be consistent, I can have it raise an error